File: src/js/4_BehavioralPatterns/1_ChainOfResponsibility/method_chain.js
class Creature {
  constructor(name, attack, defense) {
    this.name = name;
    this.attack = attack;
    this.defense = defense;
  }

  toString() {
    return `${this.name} (${this.attack}/${this.defense})`;
  }
}

class CreatureModifier
{
  constructor(creature)
  {
    this.creature = creature;
    this.next = null;
  }

  add(modifier)
  {
    if (this.next) this.next.add(modifier);
    else this.next = modifier;
  }

  handle()
  {
    if (this.next) this.next.handle();
  }
}

class NoBonusesModifier extends CreatureModifier
{
  constructor(creature)
  {
    super(creature);
  }

  handle()
  {
    console.log('No bonuses for you!');
  }
}

class DoubleAttackModifier extends CreatureModifier
{
  constructor(creature)
  {
    super(creature);
  }

  handle()
  {
    console.log(`Doubling ${this.creature.name}'s attack`);
    this.creature.attack *= 2;
    super.handle();
  }
}

class IncreaseDefenseModifier extends CreatureModifier
{
  constructor(creature)
  {
    super(creature);
  }

  handle() {
    if (this.creature.attack <= 2)
    {
      console.log(`Increasing ${this.creature.name}'s defense`);
      this.creature.defense++;
    }
    super.handle();
  }
}

let goblin = new Creature('Goblin', 1, 1);
console.log(goblin.toString());

let root = new CreatureModifier(goblin);

//root.add(new NoBonusesModifier(goblin));

root.add(new DoubleAttackModifier(goblin));
//root.add(new DoubleAttackModifier(goblin));

root.add(new IncreaseDefenseModifier(goblin));

// eventually...
root.handle();
console.log(goblin.toString());

File: src/csharp/4_BehavioralPatterns/1_ChainOfResponsibility/MethodChain.cs
using System;
using System.Security.Cryptography;
using static System.Console;

namespace DotNetDesignPatternDemos.Behavioral.ChainOfResponsibility.MethodChain
{
  public class Creature
  {
    public string Name;
    public int Attack, Defense;

    public Creature(string name, int attack, int defense)
    {
      Name = name ?? throw new ArgumentNullException(paramName: nameof(name));
      Attack = attack;
      Defense = defense;
    }

    public override string ToString()
    {
      return $"{nameof(Name)}: {Name}, {nameof(Attack)}: {Attack}, {nameof(Defense)}: {Defense}";
    }
  }

  public class CreatureModifier
  {
    protected Creature creature;
    protected CreatureModifier next;
    
    public CreatureModifier(Creature creature)
    {
      this.creature = creature ?? throw new ArgumentNullException(paramName: nameof(creature));
    }

    public void Add(CreatureModifier cm)
    {
      if (next != null) next.Add(cm);
      else next = cm;
    }

    public virtual void Handle() => next?.Handle();
  }

  public class NoBonusesModifier : CreatureModifier
  {
    public NoBonusesModifier(Creature creature) : base(creature)
    {
    }

    public override void Handle()
    {
      // nothing
      WriteLine("No bonuses for you!");
    }
  }

  public class DoubleAttackModifier : CreatureModifier
  {
    public DoubleAttackModifier(Creature creature) : base(creature)
    {
    }

    public override void Handle()
    { 
      WriteLine($"Doubling {creature.Name}'s attack");
      creature.Attack *= 2;
      base.Handle();
    }
  }

  public class IncreaseDefenseModifier : CreatureModifier
  {
    public IncreaseDefenseModifier(Creature creature) : base(creature)
    {
    }

    public override void Handle()
    {
      WriteLine("Increasing goblin's defense");
      creature.Defense += 3;
      base.Handle();
    }
  }

  public class Demo
  {
    static void Main(string[] args)
    {
      var goblin = new Creature("Goblin", 2, 2);
      WriteLine(goblin);
      
      var root = new CreatureModifier(goblin);

      root.Add(new NoBonusesModifier(goblin));

      WriteLine("Let's double goblin's attack...");
      root.Add(new DoubleAttackModifier(goblin));

      WriteLine("Let's increase goblin's defense");
      root.Add(new IncreaseDefenseModifier(goblin));

      // eventually...
      root.Handle();
      WriteLine(goblin);
    }
  }
}